Created by Thai Vang and Navpreet Kloy

Dataset from: https://www.kaggle.com/datasets/ttv9129/gamestore-dataset

In [ ]:
#Adding code to get rid of deprecation warning due to outdated imports.
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning)

In [ ]:
# call the required libraries
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules

In [ ]:
#Importing dataset into colab
df=pd.read_csv('gamestore.csv')
df.head()

,"Game Pass,Game,Batteries,Gift Card"
0,"Game,Controller,Gift Card"
1,Game
2,"Game,Headset,Game Pass"
3,"Batteries,Headset,Game Pass"
4,Console


In [ ]:
#Adding prodcut as column header
df=pd.read_csv('gamestore.csv', names=['Product'])
df.head()

,Product
0,"Game Pass,Game,Batteries,Gift Card"
1,"Game,Controller,Gift Card"
2,Game
3,"Game,Headset,Game Pass"
4,"Batteries,Headset,Game Pass"


In [ ]:
#Code to show shape of the dataset showing 400 rows with one column
df.shape

(400, 1)

In [ ]:
#converting pandas dataframe into a list of lists
data=list(df["Product"].apply(lambda x:x.split(",")))
data

[['Game Pass', 'Game', 'Batteries', 'Gift Card'],
 ['Game', 'Controller', 'Gift Card'],
 ['Game'],
 ['Game', 'Headset', 'Game Pass'],
 ['Batteries', 'Headset', 'Game Pass'],
 ['Console'],
 ['Game Pass'],
 ['Game Pass', 'Batteries', 'Headset'],
 ['Game Pass', 'Console', 'Gift Card'],
 ['Batteries', 'Game Pass'],
 ['Controller', 'Game Pass', 'Batteries'],
 ['Game Pass', 'Console', 'Game'],
 ['Headset', 'Gift Card', 'Game', 'Console'],
 ['Game Pass', 'Game', 'Gift Card'],
 ['Batteries', 'Gift Card'],
 ['Gift Card', 'Batteries', 'Controller'],
 ['Gift Card', 'Controller', 'Console'],
 ['Controller'],
 ['Game', 'Batteries', 'Gift Card', 'Console'],
 ['Headset', 'Console'],
 ['Batteries', 'Game', 'Controller', 'Gift Card'],
 ['Game'],
 ['Controller'],
 ['Game Pass'],
 ['Batteries', 'Gift Card', 'Controller'],
 ['Game'],
 ['Headset', 'Game Pass', 'Controller', 'Game'],
 ['Headset', 'Game', 'Gift Card', 'Console'],
 ['Console', 'Gift Card'],
 ['Gift Card', 'Headset', 'Batteries'],
 ['Console',

In [ ]:
print(df.isnull().sum())  # Counts the number of missing (NaN) values in each column


Product    0
dtype: int64


In [ ]:
#transform the list with one-hot encoding
from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)

#Turn False & True to 0 and 1 for readability
df=df.replace(False,0)
df=df.replace(True,1)

df

/tmp/ipython-input-2362902784.py:9: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df=df.replace(True,1)


,Batteries,Console,Controller,Game,Game Pass,Gift Card,Headset
0,1,0,0,1,1,1,0
1,0,0,1,1,0,1,0
2,0,0,0,1,0,0,0
3,0,0,0,1,1,0,1
4,1,0,0,0,1,0,1
...,...,...,...,...,...,...,...
395,0,0,0,0,1,0,0
396,0,0,1,1,0,0,1
397,0,1,0,0,0,1,1
398,0,1,1,1,1,0,0


In [ ]:
# SUPPORT threshold value
df=apriori(df,min_support=0.1,use_colnames=True)
df

,support,itemsets
0,0.3650,(Batteries)
1,0.3525,(Console)
2,0.3775,(Controller)
3,0.3600,(Game)
4,0.3225,(Game Pass)
5,0.3975,(Gift Card)
6,0.3450,(Headset)
7,0.1175,"(Batteries, Console)"
8,0.1375,"(Batteries, Controller)"
9,0.1125,"(Batteries, Game)"


In [ ]:
#Using association_rules for the confidence parameter
df_ar=association_rules(df,metric="confidence", min_threshold=0.6)
df_ar

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski


In [ ]:
#Here is the LIFT parameter.
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,(Batteries),(Gift Card),0.3650,0.3975,0.1600,0.438356,1.102783,1.0,0.014912,1.072744,0.146777,0.265560,0.067811,0.420436
1,(Gift Card),(Batteries),0.3975,0.3650,0.1600,0.402516,1.102783,1.0,0.014912,1.062789,0.154694,0.265560,0.059080,0.420436
2,(Console),(Controller),0.3525,0.3775,0.1400,0.397163,1.052088,1.0,0.006931,1.032618,0.076462,0.237288,0.031587,0.384012
3,(Controller),(Console),0.3775,0.3525,0.1400,0.370861,1.052088,1.0,0.006931,1.029184,0.079532,0.237288,0.028357,0.384012
4,(Console),(Headset),0.3525,0.3450,0.1275,0.361702,1.048412,1.0,0.005888,1.026167,0.071315,0.223684,0.025499,0.365634
5,(Headset),(Console),0.3450,0.3525,0.1275,0.369565,1.048412,1.0,0.005888,1.027069,0.070498,0.223684,0.026356,0.365634
6,(Game),(Headset),0.3600,0.3450,0.1250,0.347222,1.006441,1.0,0.000800,1.003404,0.010000,0.215517,0.003393,0.354771
7,(Headset),(Game),0.3450,0.3600,0.1250,0.362319,1.006441,1.0,0.000800,1.003636,0.009771,0.215517,0.003623,0.354771
8,(Headset),(Game Pass),0.3450,0.3225,0.1125,0.326087,1.011122,1.0,0.001238,1.005323,0.016794,0.202703,0.005294,0.337462
9,(Game Pass),(Headset),0.3225,0.3450,0.1125,0.348837,1.011122,1.0,0.001238,1.005893,0.016236,0.202703,0.005858,0.337462


In [ ]:
#Here is the CONVICTION parameter.
df_ar=association_rules(df,metric="conviction", min_threshold=1.0)
df_ar


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,(Batteries),(Gift Card),0.3650,0.3975,0.1600,0.438356,1.102783,1.0,0.014912,1.072744,0.146777,0.265560,0.067811,0.420436
1,(Gift Card),(Batteries),0.3975,0.3650,0.1600,0.402516,1.102783,1.0,0.014912,1.062789,0.154694,0.265560,0.059080,0.420436
2,(Console),(Controller),0.3525,0.3775,0.1400,0.397163,1.052088,1.0,0.006931,1.032618,0.076462,0.237288,0.031587,0.384012
3,(Controller),(Console),0.3775,0.3525,0.1400,0.370861,1.052088,1.0,0.006931,1.029184,0.079532,0.237288,0.028357,0.384012
4,(Console),(Headset),0.3525,0.3450,0.1275,0.361702,1.048412,1.0,0.005888,1.026167,0.071315,0.223684,0.025499,0.365634
5,(Headset),(Console),0.3450,0.3525,0.1275,0.369565,1.048412,1.0,0.005888,1.027069,0.070498,0.223684,0.026356,0.365634
6,(Game),(Headset),0.3600,0.3450,0.1250,0.347222,1.006441,1.0,0.000800,1.003404,0.010000,0.215517,0.003393,0.354771
7,(Headset),(Game),0.3450,0.3600,0.1250,0.362319,1.006441,1.0,0.000800,1.003636,0.009771,0.215517,0.003623,0.354771
8,(Headset),(Game Pass),0.3450,0.3225,0.1125,0.326087,1.011122,1.0,0.001238,1.005323,0.016794,0.202703,0.005294,0.337462
9,(Game Pass),(Headset),0.3225,0.3450,0.1125,0.348837,1.011122,1.0,0.001238,1.005893,0.016236,0.202703,0.005858,0.337462


The association rule analysis identified several meaningful purchasing patternsw ithin the Gamestore dataset. The strongest rule was Batteries -> Gift Card, with a support of 0.16, confidence of 0.438, and lift of 1.103. This means that 16% of all transactions contained both items, and when Batteries were purchaced, there was a 43.8% chance that a Gift Card was also included. The lieft value greater than 1 indicated a positive relationship, meaning customers who buy Batteries are about 10% more likely to also purchase a Gift Card compared to random chance. Other notable patterns include Console -> Controller (confidence around 39.7%) and Console -> Headset (confidence around 36.2%), suggesting that accessories are commonly purchased alongside consoles.

Overall, the lift values across the rules are slightly above 1, which confirms positive but moderate associations between products. While the relationships are not extremely strong, they stulle provide useful business insight. These patterns suggest opportunities for bundling consoles with controllers or headset, recommending batteries with related accessories, and strategically placing complementary products near each other in-store or online. Leveraging these associations could help increase average transaction value and improve cross-selling effectiveness.